notebooks/USE_embeddings.py

```diff
@@ -0,0 +1,220 @@
+content_store_df = pd.read_csv("/home/james/Downloads/preprocessed_content_store_210920.csv",
+                               compression='gzip', delimiter="\t", low_memory=False)
+cid_df = pd.DataFrame({'content_id': list(cid_count.keys()), 'relationship_flag': 1})
+cols_keep = ['document_type', 'content_id', 'first_published_at', 'details', 'description']
+subset_content_df = content_store_df.merge(cid_df, on='content_id', how='inner')[cols_keep].copy()
+subset_content_df['details'] = subset_content_df['details'].map(literal_eval)
+
+
+def clean_xml(original_text):
+    ''' strips out xml tagging from string'''
+    extracted_sentence = []
+    start_idx = 1
+    end_idx = 1
+    while (start_idx > 0) and (end_idx > 0):
+        end_idx = original_text.find('<')
+        if end_idx >= 0:
+            extracted_sentence.append(original_text[:end_idx])
+            start_idx = original_text.find('>')
+            if (start_idx >= 0):
+                original_text = original_text[start_idx + 1:]
+    if len(original_text) > 0:
+        extracted_sentence.append(original_text)
+    return str(''.join(extracted_sentence))
+
+
+def extract_paragraphs(original_text):
+    ''' takes raw string text from gov uk and returns extracted paragraphs
+    still contains xml tags'''
+    extracted_paragraphs = []
+    start_idx = 1
+    end_idx = 1
+    while (start_idx >= 0) and (end_idx >= 0):
+        start_idx = original_text.find('<p>')
+        end_idx = original_text.find('</p>')
+        if (start_idx >= 0) and (end_idx >= 0):
+            if (end_idx - start_idx) > 3:
+                cleaned_text_segment = clean_xml(original_text[start_idx + 3:end_idx])
+                extracted_paragraphs.append(cleaned_text_segment)
+            original_text = original_text[end_idx + 3:]
+    return extracted_paragraphs
+
+
+def document_embedding(paragraphs):
+    '''
+    average embeddings across sentences
+    '''
+    embedding = embed(paragraphs)
+    average_embedding = tf.math.reduce_mean(embedding, axis=0).numpy()
+    return average_embedding
+
+
+# initialise an empty array for embeddings
+collected_doc_embeddings = np.zeros((subset_content_df.shape[0], 512))
+
+# fill array with embeddings for all docs
+# problem with out of memory warnings on 8gb card: reduce extracted paragraphs to 128
+# flake8 complaining with the complexity of code beneath which generates embeddings using the description
+'''
+for i in range(subset_content_df.shape[0]):
+    try:
+        doc = subset_content_df.iloc[i]['details']['body']
+        if type(doc) == list:
+            doc = doc[0]['content']
+    except KeyError:
+        try:
+            doc = subset_content_df.iloc[i]['description']
+            if isNaN(doc):
+                continue
+        except KeyError:
+            continue
+    extracted_paragraphs = extract_paragraphs(doc)
+    if len(extracted_paragraphs) > 128:
+        extracted_paragraphs = extracted_paragraphs[:128]
```

Comment:
Question here about whether this is all that necessary given we are trying to clean the text and extract paragraphs. Surely the `description` or `text` columns are what we want?